games/TicTacToe/contract/src/main.sw

```diff
@@ -0,0 +1,98 @@
+
+use data_structures::{Game, Players, Winner, core::ops::Eq};
+
+storage {
+    game: Game,
+    player_turn: u64,
+    map: StorageMap<u64,
+    u64>, 
+}
+
+abi TicTacToe {
+    #[storage(write)]fn new_game(player_one: Players, player_two: Players) -> Game;
+    #[storage(read, write)]fn make_move(position: u64);
+    #[storage(read, write)]fn next_player();
+    #[storage(write)]fn end_game(game: Game) -> Winner;
+    #[storage(read)]fn map_is_full() -> bool;
+}
+
+impl TicTacToe for Contract {
+    #[storage(write)]fn new_game(player_one: Players, player_two: Players) -> Game {
+        let mut game = Game {
+            PlayerOne: player_one,
+            PlayerTwo: player_two,
+            winner: Winner::None,
+        };
+        storage.player_turn = 1;
+        let mut counter = 0;
+        while counter < 9 {
+            counter += 1;
+            insert_into_map(counter, 0)
+        }
+        return game;
+    }
+
+    // This function first checks who's turn it is and then inserts to the storage map the player key + the position
+    #[storage(read, write)]fn make_move(position: u64) {
+        if storage.player_turn == 1 {
+            insert_into_map(1, position);
+        }
+        if storage.player_turn == 2 {
+            insert_into_map(2, position);
+        }
+    }
+
+    // This function first checks who's turn it is and then switches to the other one.
+    #[storage(read, write)]fn next_player() {
+        // accessing the player_turn state in the storage and switches it to the next one
+        if storage.player_turn == 1 {
+            storage.player_turn = 2;
+        }
+        if storage.player_turn == 2 {
+            storage.player_turn = 1;
+        }
+    }
+
+    //Check each cell. If one of them is empty (contains 0), then the map isn't full yet.
+    #[storage(read)]fn map_is_full() -> bool {
+        // Acessing the state of each cell. If it's different than 0 (empty cell),
+        // then it keeps counting the number of not-empty cells.
+        // If all 9 cells are either 1 or 2, it means the map is full and the function returns true.
+        // Else it's not full and the function returns false.
+        let mut counter = 0;
+        let mut break_early = false;
+        let mut result = true;
+        while counter < 9 {
+            if break_early == true {
+                // here we ensure the condition will evaluate to false, breaking the loop
+                counter = 10;
+                result = false;
+            }
+            counter += 1;
+            if get_from_map(counter) == 0 {
+                break_early = true;
+            }
+        }
+        return result;
+    }
+
+    // save the game and return the winner
+    #[storage(write)]fn end_game(game: Game) -> Winner {
+        storage.game = game;
+        return game.winner;
+    }
+}
+
+//save the player moves in the storage
+#[storage(write)]fn insert_into_map(key: u64, value: u64) {
```

Comment:
I don't know if wrapping the Sway syntax for stroage set/get adds anything in terms of readability here.

Reply:
How else would you store the moves?

Reply:
I just mean, you can do `storage.map.insert(key, value);` directly in the function that needs to update the map, rather than abstracting the `insert` operation behind another function.